apis/mysql/v1alpha1/grant_types.go

```diff
@@ -0,0 +1,135 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+)
+
+const (
+	errUserNotFound = "referenced user not found"
+	errDBNotFound   = "referenced database not found"
+	errNoUser       = "no user specified"
+	errNoDB         = "no database specified"
+)
+
+// A GrantSpec defines the desired state of a Grant.
+type GrantSpec struct {
+	runtimev1alpha1.ResourceSpec `json:",inline"`
+	ForProvider                  GrantParameters `json:"forProvider"`
+}
+
+// GrantParameters define the desired state of a MySQL grant instance.
+type GrantParameters struct {
+	// Privileges to be granted.
+	// See https://mariadb.com/kb/en/grant/#database-privileges for available privileges.
+	Privileges []string `json:"privileges"`
+
+	// User this grant is for.
+	// +optional
+	User *string `json:"user,omitempty"`
+
+	// UserRef references the user object this grant is for.
+	// +optional
+	UserRef *runtimev1alpha1.Reference `json:"userRef,omitempty"`
+
+	// Database this grant is for.
+	// +optional
+	Database *string `json:"database,omitempty"`
+
+	// DatabaseRef references the database object this grant it for.
+	// +optional
+	DatabaseRef *runtimev1alpha1.Reference `json:"databaseRef,omitempty"`
+}
+
+// A GrantStatus represents the observed state of a Grant.
+type GrantStatus struct {
+	runtimev1alpha1.ResourceStatus `json:",inline"`
+}
+
+// +kubebuilder:object:root=true
+
+// A Grant represents the declarative state of a MySQL grant.
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="SYNCED",type="string",JSONPath=".status.conditions[?(@.type=='Synced')].status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster
+type Grant struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   GrantSpec   `json:"spec"`
+	Status GrantStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// GrantList contains a list of Grant
+type GrantList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Grant `json:"items"`
+}
+
+// GetUsername returns the referenced username
```

Comment:
We actually have a library to help implement this pattern since it's fairly common. Sorry I neglected to point this out earlier! If your type satisfies the `ReferenceResolver` interface we'll automatically resolve references. You can use `pkg/reference` from crossplane-runtime to implement that interface in a fashion pretty similar to what you have here.

https://github.com/crossplane/crossplane-runtime/blob/57ef78/pkg/reconciler/managed/reconciler.go#L147
https://github.com/crossplane/provider-aws/blob/67fbb0c1e/apis/database/v1beta1/referencers.go

Reply:
Wow, this is awesome. Makes it a lot easier.
Is it safe to just dereference the `Database` and `User` pointers without `nil` checks (since they should always be set by the reference resolvers, right)?

Reply:
>  Is it safe to just dereference the `Database` and `User` pointers without `nil` checks (since they should always be set by the reference resolvers, right)?

Ah, I see. When I configure my reconciler with `managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),` then the pointer should always be set.